src/journal/compress.c

```diff
@@ -677,12 +905,99 @@ int decompress_stream_lz4(int in, int out, uint64_t max_bytes) {
+        _cleanup_free_ char *out = malloc(out_size + in_size);
+        char *in = out + out_size;
+        ZSTD_outBuffer ZSTD_out = {out, out_size, 0};
+        ZSTD_inBuffer ZSTD_in = {in, in_size, 0};
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        if (!s || !out)
+                return -ENOMEM;
+
+        ret = ZSTD_initDStream(s);
+        if (ZSTD_isError(ret)) {
+                log_debug("Failed to initialize ZSTD encoder: %s", ZSTD_getErrorName(ret));
+                return zstd_ret_to_errnoh(ret);
+        }
+
+        for (;;) {
+                bool first = true;
+
+                ss = read(fdf, in, in_size);
+                if (ss < 0)
+                        return -errno;
+                if (ss == 0)
+                        break;
+                ZSTD_in.size = ss;
+                in_bytes += ss;
+
+                if (first) {
+                        unsigned long long q;
+
+                        q = ZSTD_getFrameContentSize(ZSTD_in.src, ZSTD_in.size);
+                        if (q == ZSTD_CONTENTSIZE_UNKNOWN)
+                                ;
+                        else if (q == ZSTD_CONTENTSIZE_ERROR)
+                                return -EBADMSG;
+                        else if (q > max_bytes)
+                                return -EFBIG;
+                        first = false;
+                }
+
+                while (ZSTD_in.pos < ZSTD_in.size) {
+                        ret = ZSTD_decompressStream(s, &ZSTD_out, &ZSTD_in);
+                        if (ZSTD_isError(ret)) {
+                                log_debug("ZSTD dencoder failed: %s", ZSTD_getErrorName(ret));
+                                return zstd_ret_to_errnoh(ret);
+                        }
+
+                        if (ZSTD_in.pos == ZSTD_in.size)
+                                break;
+                }
+
+                if (left < ZSTD_out.pos)
+                        return -EFBIG;
+
+                r = loop_write(fdt, ZSTD_out.dst, ZSTD_out.pos, false);
+                if (r < 0)
+                        break; /* cleanup and return error */
+
+                left -= ZSTD_out.pos;
+                ZSTD_out.pos = 0;
+                if (ZSTD_in.pos == ZSTD_in.size)
+                        break;
+        }
+
+        log_debug("ZSTD decompression finished (%"PRIu64" -> %"PRIu64" bytes, %.1f%%)",
+                in_bytes, max_bytes - left,
+                (double) (max_bytes - left) / in_bytes * 100);
+        return 0;
+#else
+        log_debug("Cannot decompress file. Compiled without ZSTD support.");
+        return -EPROTONOSUPPORT;
+#endif
+}
+
 int decompress_stream(const char *filename, int fdf, int fdt, uint64_t max_bytes) {
 
         if (endswith(filename, ".lz4"))
                 return decompress_stream_lz4(fdf, fdt, max_bytes);
         else if (endswith(filename, ".xz"))
                 return decompress_stream_xz(fdf, fdt, max_bytes);
+        else if (endswith(filename, ".zst"))
```

Comment:
further up you use ".zstd" as suffix, which one is the correct one?

Reply:
It looks like the code now agrees on `.zst`, which is what Zstd references as the official extension: https://github.com/facebook/zstd/wiki